contracts/swaps/connectors/SwapsImplSovrynSwap.sol

```diff
@@ -0,0 +1,176 @@
+    {
+        require(sourceTokenAddress != destTokenAddress, "source == dest");
+        require(supportedTokens[sourceTokenAddress] && supportedTokens[destTokenAddress], "invalid tokens");
+
+        ISovrynSwapNetwork sovrynSwapNetwork = getSovrynSwapNetworkContract();
+        IERC20[] memory path = sovrynSwapNetwork.conversionPath(
+            IERC20(sourceTokenAddress),
+            IERC20(destTokenAddress)
+        );
+
+        uint expectedReturn = 0;
+
+        //if the required amount of destination tokens is passed, we need to calculate the estimated amount of source tokens
+        //regardless of the minimum source token amount (name is misleading)
+        if(requiredDestTokenAmount > 0){
+            sourceTokenAmountUsed = estimateSourceTokenAmount(sourceTokenAddress, destTokenAddress, requiredDestTokenAmount,  maxSourceTokenAmount);
+             //sovrynSwapNetwork.rateByPath does not return a rate, but instead the amount of destination tokens returned
+            require(sovrynSwapNetwork.rateByPath(path, sourceTokenAmountUsed) >= requiredDestTokenAmount, "insufficient source tokens provided.");
+            expectedReturn = requiredDestTokenAmount;
+        }
+        else if(minSourceTokenAmount > 0){
+            sourceTokenAmountUsed = minSourceTokenAmount;
+            expectedReturn = sovrynSwapNetwork.rateByPath(path, minSourceTokenAmount);
+        }
+        
+        require(sourceTokenAmountUsed > 0, "cannot swap 0 tokens");
+        
+        allowTransfer(sourceTokenAmountUsed, sourceTokenAddress, address(sovrynSwapNetwork));
+
+        //note: the kyber connector uses .call() to interact with kyber to avoid bubbling up. here we allow bubbling up.
+        destTokenAmountReceived = sovrynSwapNetwork.convertByPath(path, sourceTokenAmountUsed, expectedReturn, address(0), address(0), 0);
+        
+        //if the sender is not the protocol (calling with delegatecall), return the remainder to the specified address.
+        //note: for the case that the swap is used without the protocol. not sure if it should, though. needs to be discussed.
+        if (returnToSenderAddress != address(this)) {
+            if (sourceTokenAmountUsed < maxSourceTokenAmount) {
+                // send unused source token back
+                IERC20(sourceTokenAddress).safeTransfer(
+                    returnToSenderAddress,
+                    maxSourceTokenAmount-sourceTokenAmountUsed
+                );
+            }
+        }
+
+    }
+
+    /**
+     * check is the existing allowance suffices to transfer the needed amount of tokens.
+     * if not, allows the transfer of an arbitrary amount of tokens.
+     * @param tokenAmount the amount to transfer
+     * @param tokenAddress the address of the token to transfer
+     * @param sovrynSwapNetwork the address of the sovrynSwap network contract.
+     * */
+    function allowTransfer(uint256 tokenAmount, address tokenAddress, address sovrynSwapNetwork) internal{
```

Comment:
I prefer this way to style the code. However, in order to maintain one style I would recommend using line 45 style

Reply:
I find it much better readable, too. Feel free to adapt it.